Add a plugin tutorial

choldgraf opened the pull request:
This adds a tutorial for creating basic plugins with Jupyter Book. It covers creating and registering plugins, creating a directive, and creating a transform.

It also cleans up some navigation and adds two how-tos related to plugins.

This is a result of some self-learning I had to do in creating some plugins for my website, so hopefully this makes things easier to learn for future readers!

### Follow ups

A few things I opened along the way:

- https://github.com/jupyter-book/mystmd/issues/1651
- https://github.com/jupyter-book/mystmd/issues/1649
- https://github.com/jupyter-book/mystmd/issues/1652
- https://github.com/jupyter-book/mystmd/issues/1654
- https://github.com/jupyter-book/mystmd/issues/1655

choldgraf commented on docs/glossary.md:
```diff
@@ -20,4 +20,10 @@ Environment Manager
+: A way of structuring and labeling documents according to a specification to ensure consistent behavior. The MyST document engine [defines its own document specification](https://mystmd.org/spec), and all MyST documents are first built into ASTs before being rendered into output formats (like HTML, PDF, etc).
+
+AST
+: See {term}`Abstract Syntax Tree`.
```
Opened a new issue with a nicer UX here:

- https://github.com/jupyter-book/mystmd/issues/1649

fperez commented:
Awesome, thx @choldgraf!! Pinging both @minrk and @ryanlovett who have been thinking about this as well, they may have input/suggestions.

ryanlovett commented on docs/plugins/debug.md:
```diff
@@ -0,0 +1,62 @@
+# Debug plugins
+
+This section describes a few days to debug MyST plugins.
```
I believe you mean "ways".

agoose77 replied:
```suggestion
This section describes a few ways to debug MyST plugins.
```

agoose77 commented on docs/tutorial/plugins.md:
````diff
@@ -0,0 +1,547 @@
+🛠 Next initialize a Jupyter Book in the folder to add a configuration file:
+
+```shell
+$ jupyter-book init
````
In JB, we are currently leaning in to the `jupyter book` dispatch mechanism, so we should change this.

agoose77 commented on docs/tutorial/plugins.md:
````diff
@@ -0,0 +1,547 @@
+
+🛠 Finally, create an empty file that we'll use to play around with our plugin:
+
+```shell
+$ echo '# My demo page' > myplugindemo.md
+```
````
````suggestion
```{code} markdown
:filename: myplugindemo.md

# My demo page
```
````

I think prefer platform agnostic instructions?

agoose77 commented on docs/tutorial/plugins.md:
````diff
@@ -0,0 +1,547 @@
+    - file: myplugindemo.md
+```
+
+You should now be able to run `myst start` and see a page with just your title rendered:
````
```suggestion
You should now be able to run `jupyter book start` and see a page with just your title rendered:
```

agoose77 commented on docs/tutorial/plugins.md:
```diff
@@ -0,0 +1,547 @@
+
+You should now be able to run `myst start` and see a page with just your title rendered:
+
+🛠️ Run `myst start` to confirm that your demo book is set up properly.
```
```suggestion
🛠️ Run `jupyter book start` to confirm that your demo book is set up properly.
```

agoose77 commented on docs/tutorial/plugins.md:
````diff
@@ -0,0 +1,547 @@
+🛠️ Run `myst start` to confirm that your demo book is set up properly.
+
+```shell
+$ myst start
````
```suggestion
$ jupyter book start
```

agoose77 commented on docs/tutorial/plugins.md:
```diff
@@ -0,0 +1,547 @@
+
+Next we'll create an empty JavaScript file[^esm] that we'll use to add new plugin logic.
+
+[^esm]: The format of the Javascript should be an ECMAScript modules, not CommonJS. This means it uses `import` statements rather than `require()` and is the most modern style of Javascript. See [the MDN modules documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) for background on modules.
```
```suggestion
[^esm]: The format of the JavaScript should be an ECMAScript modules, not CommonJS. This means it uses `import` statements rather than `require()` and is the most modern style of JavaScript. See [the MDN modules documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) for background on modules.
```

agoose77 commented on docs/tutorial/plugins.md:
````diff
@@ -0,0 +1,547 @@
+🛠 Start your MyST site.
+
+```shell
+$ myst start
````
```suggestion
$ jupyter book start
```

agoose77 commented on docs/tutorial/plugins.md:
````diff
@@ -0,0 +1,547 @@
+
+🛠 Create a `src/` folder and add an empty JavaScript file to it
+
+```shell
+$ mkdir src
+$ touch src/myplugin.mjs
+```
````
```suggestion
```{code}
:filename: src/myplugin.mjs

```
````

agoose77 commented on docs/tutorial/plugins.md:
````diff
@@ -0,0 +1,547 @@
+🛠️ **Re-build** MyST (you must re-build manually any time you change a plugin)
+
+```shell
+$ myst start
````
```suggestion
$ jupyter book start
```

agoose77 commented on docs/tutorial/plugins.md:
````diff
@@ -0,0 +1,547 @@
+🛠️ Re-start MyST to confirm that the plugin's transform is loaded.
+
+```shell
+$ myst start
````
```suggestion
$ jupyter book start
```

agoose77 commented on docs/tutorial/plugins.md:
````diff
@@ -0,0 +1,547 @@
+🛠️ Re-build your demo site.
+
+```shell
+$ myst start
````
```suggestion
$ jupyter book start
```

agoose77 commented on docs/plugins/sandbox.md:
```diff
@@ -0,0 +1,48 @@
+# Use the MyST Sandbox to identify card AST structure
+
+The [MyST interactive sandbox](https://mystmd.org/sandbox) is a great way to explore what MyST looks like when it is rendered, and what its unerlying AST structure looks like.
```
```suggestion
The [MyST interactive sandbox](https://mystmd.org/sandbox) is a great way to explore what MyST looks like when it is rendered, and what its underlying AST structure looks like.
```

agoose77 commented on docs/plugins/sandbox.md:
```diff
@@ -0,0 +1,48 @@
+
+This is particularly useful if you're generating MyST AST from scratch. For example, as part of a [plugin role or directive](../tutorial/plugins.md).
+
+For example, here's a video that shows how to use the MyST sandbox to explore the structure of a `{card}` directive.
```
```suggestion
For example, here's a video that shows how to use the MyST sandbox to explore the structure of a {myst:directive}`card` directive.
```

agoose77 commented on docs/tutorial/website.md:
````diff
@@ -33,7 +33,7 @@ Through the course of the tutorials we will add content to these documents that
 🛠 Download the example content[^no-git], and navigate into the folder:
 
-```bash
+```shell
 git clone https://github.com/jupyter-book/mystmd-quickstart.git
 cd mystmd-quickstart
 ```
````
```suggestion
```shell
$ git clone https://github.com/jupyter-book/mystmd-quickstart.git
$ cd mystmd-quickstart
```
```

agoose77 commented on docs/plugins/sandbox.md:
````diff
@@ -0,0 +1,48 @@
+For example, here's a video that shows how to use the MyST sandbox to explore the structure of a {myst:directive}`card` directive.
+
+```{figure} media/sandbox-demo.mp4
+Using [the MyST sandbox](https://mystmd.org/sandbox) to preview and show the AST of the `{card}` directive.
````
```suggestion
Using [the MyST sandbox](https://mystmd.org/sandbox) to preview and show the AST of the {myst:directive}`card` directive.
```

choldgraf commented:
I added a brief update to show off how to use the `ctx` command and put it in a howto as well. @agoose77 feel free to take a pass and merge whenever you like!

agoose77 commented on docs/plugins/ast.md:
```diff
@@ -0,0 +1,117 @@
+`ctx.parseMyst` returns a `root` node, which contains the parsed MyST.
+This function takes the first child of the root node, which is usually better for inserting into the document.
+
+You can also use this to manually construct MyST outputs using the arguments of your directive as data. For example:
```
Still not 100% sure on encouraging use of `parseMyst` for constant nodes. @rowanc1 any thoughts on the implications of this?

agoose77 replied:
If we don't hear back, I will merge later today.

choldgraf replied:
What are your concerns with the workflow?

Alternatively we could remove the use of this from the tutorial (and revert to hand crafted AST), and out a big warning on the howto guide based on your concerns.

agoose77 commented on docs/plugins/ast.md:
````diff
@@ -0,0 +1,117 @@
+
+:::{tip} If you need to use multi-line strings you must dedent them
+The example above puts a multi-line string onto one line by manually coding the `\n` characters.
+If you instead want to show a multi-line string in your code, for example like the following:
+
+
+```{code} javascript
+:filename: src/justacard.mjs
+const myDirective = {
+  name: "justacard",
+  doc: "Basically does what a card directive does!",
+  arg: { type: String, doc: "To be parsed into MyST AST." },
+  body: { type: String, doc: "The body of the directive." },
+  run(data, vfile, ctx) {
+    const ast = ctx.parseMyst(`
+    :::{card} ${data.arg}
+    ${data.body}
+    :::`);
+    return ast.children[0];
+  },
+};
+```
+
+You'll need to dedent the string manually to get the multi-line string to work.
+The easiest way to do this is to use the [`dedent` JavaScript package](https://www.npmjs.com/package/dedent).
````
````suggestion
If you instead want to show a multi-line string in your code, you will need to remove the indentation manually, for example like the following:


```{code} javascript
:filename: src/justacard.mjs
const myDirective = {
  name: "justacard",
  doc: "Basically does what a card directive does!",
  arg: { type: String, doc: "To be parsed into MyST AST." },
  body: { type: String, doc: "The body of the directive." },
  run(data, vfile, ctx) {
    const ast = ctx.parseMyst(`
:::{card} ${data.arg}
${data.body}
::: `);
    return ast.children[0];
  },
};
```
````

I don't think we need to encourage using another package to work around a style preference -- zero indentation literals are fairly standard practice.

agoose77 commented on docs/tutorial/plugins.md:
```diff
@@ -138,8 +139,10 @@ Here's how to interpret the above:
+- **`run(data, vfile, ctx)`**: A special function all directives must have. This is executed when the directive is run.
+  - **`data`**: A data object provided to each directive when it is run. This contains information about the directive, and also other information about what's on the page.
+  - **`vfile`**: A [VFile](https://github.com/vfile/vfile) object. This is a virtual file object that contains information about the page that the directive is being run on.
+  - **`ctx`**: A context object provided to each directive when it is run. This provides helper functions and information for working with MyST.
```
Tempted to use a deflist here.

choldgraf commented:
I've added another little how-to that has boilerplate for how to create a role, directive, and transform. It doesn't really add any new content, just gives quick boilerplate users can copy into their own plugins.

I also added a warning to make clear that `ctx.parseMyst` is experimental and might change.

I'll plan to merge this one in and we can iterate on the content from there - I'm happy to make changes if folks spot something we shouldn't have put up there!
